sdks/go/pkg/beam/runners/prism/internal/jobservices/management.go

```diff
@@ -243,21 +243,21 @@ func (s *Server) Run(ctx context.Context, req *jobpb.RunJobRequest) (*jobpb.RunJ
 // Otherwise, returns nil if Job does not exist or the Job's existing state as part of the CancelJobResponse.
 func (s *Server) Cancel(_ context.Context, req *jobpb.CancelJobRequest) (*jobpb.CancelJobResponse, error) {
 	s.mu.Lock()
+	defer s.mu.Unlock()
 	job, ok := s.jobs[req.GetJobId()]
-	s.mu.Unlock()
 	if !ok {
 		return nil, nil
 	}
 	state := job.state.Load().(jobpb.JobState_Enum)
 	switch state {
-	case jobpb.JobState_CANCELLED, jobpb.JobState_DONE, jobpb.JobState_DRAINED, jobpb.JobState_UPDATED, jobpb.JobState_FAILED:
+	case jobpb.JobState_CANCELLED, jobpb.JobState_DONE, jobpb.JobState_DRAINED, jobpb.JobState_UPDATED, jobpb.JobState_FAILED, jobpb.JobState_STOPPED:
```

Comment:
STOPPED is unfortunately not a Terminal State, it's a badly named state for a job "that has yet to start". Essentially, jobs are initially in "stopped", after being submitted to the runner, and move to the running state once data processing can begin.

See https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/model/jobmanagement_v1/beam_job_api.pb.go#L111

A job is cancelable if it's not already in a terminal state. STOPPED is not terminal.

Was this added to resolve a bug or similar?

Reply:
I previously misunderstood the order and reason of the Prepare, Run lifecycle and their relationship with Job state. Consequently, I misunderstood the purposes of the noJobsCheck, postPrepCheck, and postRunCheck. Now that I do, I refactored the tests and got expected results as well as uncover errors in Cancel with respect to these lifecycles.